network-config-analyzer/NetworkConfigQuery.py

```diff
@@ -510,6 +511,107 @@ def exec(self):
+            for pod2 in peers_to_compare if pod1 in captured_pods else captured_pods:
+                if pod1 == pod2:
+                    continue
+                _, new_conns, _ = self.config1.allowed_connections(pod1, pod2)
+                _, old_conns, _ = self.config2.allowed_connections(pod1, pod2)
+                if new_conns != old_conns:
+                    all_diffs.append(SemanticDiffQuery.SingleDiff(pod1, pod2, old_conns-new_conns, new_conns-old_conns))
+
+        if len(all_diffs) > 0:
+            # Initialized with the 3 protocols supported by k8s
+            # This implementation is not suitable for Calico!
+            added = {}
+            removed = {}
+            for protocol in [6, 17, 132]:
+                added[ConnectionSet.protocol_number_to_name(protocol)] = {}
+                removed[ConnectionSet.protocol_number_to_name(protocol)] = {}
+            added['All connections'] = []
+            removed['All connections'] = []
+            # Hash diffs: protocol-> port ranges-> list of (from endpoint, to endpoint) tuples
+            for entry in all_diffs:
+                if entry.added.allow_all:
+                    added['All connections'].append((entry.from_ep, entry.to_ep))
+                else:
+                    for protocol in entry.added.allowed_protocols:
+                        if not ConnectionSet.protocol_supports_ports(protocol):
+                            continue
+                        protocol_name = ConnectionSet.protocol_number_to_name(protocol)
+                        port_range = str(entry.added.allowed_protocols[protocol])
+                        if port_range not in added[protocol_name]:
+                            added[protocol_name][port_range] = []
+                        added[protocol_name][port_range].append((entry.from_ep, entry.to_ep))
+                if entry.removed.allow_all:
+                    removed['All connections'].append((entry.from_ep, entry.to_ep))
+                else:
+                    for protocol in entry.removed.allowed_protocols:
+                        if not ConnectionSet.protocol_supports_ports(protocol):
+                            continue
+                        protocol_name = ConnectionSet.protocol_number_to_name(protocol)
+                        port_range = str(entry.removed.allowed_protocols[protocol])
+                        if port_range not in removed[protocol_name]:
+                            removed[protocol_name][port_range] = []
+                        removed[protocol_name][port_range].append((entry.from_ep, entry.to_ep))
+            explanation = f'Added connections:\n{self.pretty_print_diff(added)}\n'
+            explanation += f'Removed connections:\n{self.pretty_print_diff(removed)}\n'
+
+            return QueryAnswer(False, 'Base branch and analyzed commit are not semantically equivalent.', explanation)
+        return QueryAnswer(True, 'Base branch and analyzed commit are semantically equivalent.')
```

Comment:
These are GitSecure-specific messages. Should be replaced with something more generic

Reply:
done